server/src/main/java/org/opensearch/transport/ConnectionManager.java

```diff
@@ -63,6 +64,12 @@ void connectToNode(
 
     boolean nodeConnected(DiscoveryNode node);
 
+    Set<DiscoveryNode> getNodesLeftInProgress();
+
+    boolean markPendingLeftCompleted(DiscoveryNode node);
+
+    void markPendingLefts(List<DiscoveryNode> nodes);
```

Comment:
I think we can add one method to interface `markNodeForDisconnect(node)`. Others should be mostly internal.